x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -133,9 +137,69 @@ private static Decision debugYesAllowed(RoutingAllocation allocation, List<Strin
+        final var desiredNodesPreferredTier = getPreferredTierFromDesiredNodes(prioritizedTiers, nodes, desiredNodes);
+
+        if (desiredNodesPreferredTier.isPresent()) {
+            return desiredNodesPreferredTier;
+        }
+
+        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, nodes);
+    }
+
+    /**
+     * Given a list of tiers in descending order, return the tier that's present
+     * in the desired nodes with the highest priority, if none is present returns an
+     * {@code Optional.empty()}.
+     */
+    public static Optional<String> getPreferredTierFromDesiredNodes(
+        List<String> prioritizedTiers,
+        DiscoveryNodes discoveryNodes,
+        DesiredNodes desiredNodes
+    ) {
+        if (desiredNodes == null) {
+            return Optional.empty();
+        }
+
+        for (int i = 0; i < prioritizedTiers.size(); i++) {
+            final var tier = prioritizedTiers.get(i);
+            final var nextTier = i + 1 == prioritizedTiers.size() ? null : prioritizedTiers.get(i + 1);
+            if (tierNodesPresent(tier, desiredNodes.actualized())
+                || isDesiredNodeWithinTierJoining(tier, discoveryNodes, desiredNodes)
+                || nextTierIsGrowingAndCurrentTierCanHoldTheIndex(tier, nextTier, discoveryNodes, desiredNodes)) {
+                return Optional.of(tier);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static boolean nextTierIsGrowingAndCurrentTierCanHoldTheIndex(
+        String tier,
+        @Nullable String nextTier,
+        DiscoveryNodes discoveryNodes,
+        DesiredNodes desiredNodes
+    ) {
+        assert tierNodesPresent(tier, desiredNodes.actualized()) == false;
+        // If there's a plan to grow the next preferred tier, and it hasn't materialized yet,
+        // wait until all the nodes in the next tier have joined. This would avoid overwhelming
+        // the next tier if within the same plan one tier is removed and the next preferred tier
+        // grows.
+        return nextTier != null && tierNodesPresent(tier, discoveryNodes) && tierNodesPresent(nextTier, desiredNodes.pending());
+    }
+
+    private static boolean isDesiredNodeWithinTierJoining(String tier, DiscoveryNodes discoveryNodes, DesiredNodes desiredNodes) {
+        assert tierNodesPresent(tier, desiredNodes.actualized()) == false;
+        // Take into account the case when the desired nodes have been updated and the node in the tier would be replaced by
+        // a new one. In that case the desired node in the tier won't be actualized as it has to join, but we still need to ensure
+        // that at least one cluster member has the requested tier as we would prefer to minimize the shard movements in these cases.
+        return tierNodesPresent(tier, desiredNodes.pending()) && tierNodesPresent(tier, discoveryNodes);
```

Comment:
I think the terminology (pending and actualized) works well here, much easier to read now.